Validate that rewrite constraints match the requirements

danakj opened the pull request:
When doing impl lookup from a concrete value, or converting between facet values (which is implemented as impl lookup), ensure that the rewrite constraints known about the source value cover all of the requirements from the target facet type, and do not contradict any requirements.

To do this, we need additional information returned from the second stage of `LookupImplWitness()` (the stage on the other side of eval), to know the FacetType from which a symbolic `LookupImplWitness` finds its answer. The instruction itself erases that information, since it will continuing looking for a concrete answer on further (more specific) evaluations, but LookupImplWitness() needs the facet type to combine the rewrite constraints across all found witnesses and verify that they satisfy the query's constraints.

danakj committed:
validate-rewrites

danakj commented:
The changes to eval are kind of terrible, but this is a starting point that works and shows what is needed, from which we can discuss what to actually do.

zygoloid commented on toolchain/check/eval.cpp:
```diff
@@ -1895,6 +1898,58 @@ static auto IsPeriodSelf(EvalContext& eval_context, SemIR::ConstantId const_id)
+
+  return ConvertEvalResultToConstantId(
+      eval_context.context(),
+      EvalConstantInst(eval_context.context(), inst_id,
+                       inst.As<SemIR::LookupImplWitness>()),
```
I'm a bit confused here:  it looks like you deleted the function that you're calling here. Does this link?

danakj replied:
It linked, but yes this was indeed - it was dead code though so I guess I got away with it. The next commits restored that function and made this call into not dead code.

danakj replied:
FWIW one thing I considered was adding another flavour of EvalConstantInst which takes additional information (the current context specific id, would work) to avoid this eval specialization.

danakj committed:
eval

danakj committed:
todo

danakj committed:
fixtest

jonmeow commented on toolchain/check/impl_lookup.cpp:
```diff
@@ -628,16 +680,16 @@ auto EvalLookupSingleImplWitness(Context& context, SemIR::LocId loc_id,
-  // final impls only in that case. Note as in the CHECK above, the query can
-  // not be concrete in this case, so only final impls can produce a concrete
-  // witness for this query.
+  // If the self value is a (symbolic) facet value that has a symbolic
+  // witness, then we don't need to do impl lookup, except that we want to
+  // find any final impls to return a concrete witness if possible. So we
+  // limit the query to final impls only in that case. Note as in the CHECK
+  // above, the query can not be concrete in this case, so only final impls
+  // can produce a concrete witness for this query.
```
Is it worth avoiding rewrites of comments like this, to preserve git history? (this whole file has changes -- it looks like indent was changed, text was rewrapped, then indent changes were undone but rewrapping remained)

danakj replied:
This happens to me too much :/ I will fix it, it's not intended.

danakj committed:
put-comment-wrapping-back

danakj committed:
share-more-code

danakj commented:
I have gone ahead and cleaned up the impl lookup code now since it is not immediately obvious that this is _not_ how we need to/should solve this problem. I resolved a TODO to share more code between impl lookup and eval by doing the canonicalization of the `self` in `EvalLookupSingleImplWitness` and having it return that. And I added a bunch more comments, and improved some of the flow around both called to `EvalLookupSingleImplWitness`.

danakj committed:
improve-comment

danakj commented on toolchain/check/facet_type.cpp:
```diff
@@ -263,4 +265,415 @@ auto AllocateFacetTypeImplWitness(Context& context,
+  }
+
+  // Returns each RewriteRule found in the FacetType, then returns nullopt.
+  auto Next() -> std::optional<RewriteRule> {
```
I should explain what is going on here.

I wrote the tool to get rewrite rules from a facet type (here) and from a witness table (below) each as a generator object. Or you could think of it as an iterator in the shape of Rust. The reason I wrote it this way is to avoid allocating 2 more vectors on the heap just to collect things to iterate over and immediately discard/free. It's not significantly more difficult IMO to follow what is going on, the difference is just largely that it stores the current iterator as a field in each generator/iterator class.

If you're interested, in the future, we could introduce an adaptor that takes a class with an `auto Next() -> optional<T>` method and turns it into [a C++  range with begin/end](https://github.com/chromium/subspace/blob/9b6c1e1ad145cc35529569b6fce20e77cc5bda56/sus/iter/compat_ranges.h#L195-L221) that return [types with appropriate operators](https://github.com/chromium/subspace/blob/9b6c1e1ad145cc35529569b6fce20e77cc5bda56/sus/iter/__private/range_begin.h#L26-L54). That would allow us to write below
```cc
for (auto rule: MakeIter(RewriteRulesFromFacetType(...))) { ... }
```
instead of the current
```cc
for (auto rules = RewriteRulesFromFacetType(...);
     auto rule = rules.Next();) { ... }
```

danakj committed:
AddInstWithoutEval

danakj commented:
I have added `AddInstWithoutEval` which uses `AddImportedInst` to avoid running eval for the instruction being added, and used that for the `LookupImplWitness`. Open to renaming or adjusting anything of course, but wasn't sure what to rename `AddImportedInst` to so I left it for the moment.

This will benefit from https://github.com/carbon-language/carbon-lang/pull/5517 but it's not dependent on it.

This removes the eval.cpp specialization for `LookupImplWitness`.

danakj commented:
I have rebased this onto trunk to make it easier to review, and get tests to run again, so it now already includes #5517.

zygoloid commented on toolchain/check/eval.cpp:
```diff
@@ -207,6 +208,8 @@ class EvalContext {
   auto emitter() -> DiagnosticEmitterBase& { return context().emitter(); }
 
+  auto has_specific() const -> bool { return specific_id_.has_value(); }
+
```
This appears to have ended up being unused.

danakj replied:
Right, done

zygoloid commented on toolchain/check/eval.cpp:
```diff
@@ -15,6 +15,7 @@
 #include "toolchain/check/eval_inst.h"
 #include "toolchain/check/facet_type.h"
 #include "toolchain/check/generic.h"
+#include "toolchain/check/impl_lookup.h"
```
This appears to have ended up being unused.

danakj replied:
Done

zygoloid commented on toolchain/check/eval_inst.cpp:
```diff
@@ -200,26 +200,37 @@ auto EvalConstantInst(Context& /*context*/, SemIR::FunctionDecl inst)
-
-  auto result = EvalLookupSingleImplWitness(
-      context, SemIR::LocId(inst_id), inst, non_canonical_query_self_inst_id,
+  // NOTE: Do not retain a reference to the SpecificInterface obtained from a
+  // value store by SpecificInterfaceId. Doing impl lookup does deduce which can
+  // do more impl lookups, and impl lookup can add a new SpecificInterface to
+  // the store which can reallocate and invalidate any references held here into
+  // the store.
+  auto query_specific_interface =
+      context.specific_interfaces().Get(inst.query_specific_interface_id);
```
(Not in this PR:) A `SpecificInterface` is a tiny struct, and `specific_interfaces()` is a `CanonicalValueStore` so shouldn't be providing mutable access to it. Perhaps we could make its `Get` return the `SpecificInterface` by value, instead of worrying about invalidation?

danakj replied:
This concern goes away after https://github.com/carbon-language/carbon-lang/pull/5576 and we can clean up/remove such comments and use references or not based on performance expectations.

zygoloid commented on toolchain/check/eval_inst.cpp:
```diff
@@ -200,26 +200,37 @@ auto EvalConstantInst(Context& /*context*/, SemIR::FunctionDecl inst)
+  auto query_specific_interface =
+      context.specific_interfaces().Get(inst.query_specific_interface_id);
+
+  // We use the potentially non-canonical instruction substituted into the
+  // LookupImplWitness as it may be a concrete `FacetValue` that contains a
+  // concrete witness.
```
I found this comment confusing: `inst.query_self_inst_id` here has already been canonicalized, in the sense that it is the canonical representation of the value that it represents. I think the point here is that it's not necessarily a *type* value, rather than it not necessarily being canonical. (Put another way, it's not a suitable operand for a canonical `LookupImplWitness` instruction, despite being a canonical `InstId`.)

zygoloid commented on toolchain/check/facet_type.cpp:
```diff
@@ -263,4 +265,415 @@ auto AllocateFacetTypeImplWitness(Context& context,
+  }
+
+  // Maps to an index in both of vectors below. Constants with the same index
+  // value are considered equivalent.
```
Comment seems to be out of date ("both of vectors"), I guess you combined the vectors into a vector of structs?

danakj replied:
Right yeah, done.

zygoloid commented on toolchain/check/facet_type.cpp:
```diff
@@ -263,4 +265,415 @@ auto AllocateFacetTypeImplWitness(Context& context,
+      }
+      into.const_value_id_ = from.const_value_id_;
+    }
+    indexes_.ForEach([=](RewriteRule::AssociatedConstant& /*key*/, int& index) {
+      if (index == from.index) {
+        index = into.index;
+      }
+    });
```
It looks like this class is implementing a disjoint set forest (union find). This loop over all the indexes seems to make the overall algorithm asymptotically less efficient than [the optimal approach from the literature](https://en.wikipedia.org/wiki/Disjoint-set_data_structure); if we still need this, maybe we should be using something more like that?

zygoloid commented on toolchain/check/facet_type.cpp:
```diff
@@ -263,4 +265,415 @@ auto AllocateFacetTypeImplWitness(Context& context,
+          }
+        } else if (lhs_value) {
+          // Add the rhs to the lhs equivalency set.
+          AddEquivalent(*lhs_value, rhs);
```
This looks suspicious to me. Rewrite rules are one-directional; we shouldn't be rewriting the LHS to the RHS ever. In fact, I'm not sure this equivalence class machinery is what we want, for that reason: if we have `T:! I where .A = .B and .B = .A`, that should be an error: there's a cycle in the rewrites, so we don't have a canonical rewritten form for `T.A` / `T.B`.

danakj replied:
Right, I agree that should be an error though not in the verification that those rewrites satisfy the facet it's converting to. It should be an error at the construction of that facet type directly. This is something I was going to start on after this PR.

This here would never be used I think in that case, so it can go away.

zygoloid commented on toolchain/check/facet_type.cpp:
```diff
@@ -263,4 +265,415 @@ auto AllocateFacetTypeImplWitness(Context& context,
+        if (it_ == end_) {
+          return std::nullopt;
+        }
+      }
```
I think the `it_ == end_` check should be before you read through `it_` rather than after, because in the outer loop `it_` is referring to the next constraint, not the current one.

With that changed, I think this can be simplified to:
```suggestion
      if (lhs_id == SemIR::ErrorInst::InstId ||
          rhs_id == SemIR::ErrorInst::InstId) {
        continue;
      }
```

danakj replied:
Oh, yes thanks good eye.

zygoloid commented on toolchain/check/facet_type.cpp:
```diff
@@ -263,4 +265,415 @@ auto AllocateFacetTypeImplWitness(Context& context,
+      }
+
+      auto lhs_witness_access =
+          context_.insts().GetAs<SemIR::ImplWitnessAccess>(lhs_id);
```
Is it possible that the LHS is something else here? In particular, if this facet type was formed when resolving a specific, could the impl witness have already succeeded and produced a concrete constant here?

zygoloid commented on toolchain/check/facet_type.cpp:
```diff
@@ -263,4 +265,415 @@ auto AllocateFacetTypeImplWitness(Context& context,
+        auto rhs = TryGetAssociatedConstantFromImplWitnessAccess(
+            context_, *rhs_witness_access);
+        if (!rhs) {
+          // The RHS is comes from a witness table but is not an associated
+          // constant, such as if we find an error. Continue to the next element
+          // in this table.
+          continue;
```
Could the RHS be something like the name of an associated function here? Eg,
```carbon
interface I {
  fn F();
  let T:! type;
  let V:! T;
}
fn G[X:! I where .V = .F]() {}
```

danakj replied:
That's a good testcase, thanks. Added (it already passes). But if you assign the wrong function it also passes but shouldn't. I added that also:
```carbon
fn F(T:! I where .Y = .A) {}

fn G(T:! I where .Y = .B) {
  // TODO: This should be diagnosed as the wrong value being assigned to .Y for
  // what F is expecting.
  //
  F(T);
}
```

zygoloid commented on toolchain/check/facet_type.cpp:
```diff
@@ -263,4 +265,415 @@ auto AllocateFacetTypeImplWitness(Context& context,
+  llvm::ArrayRef<SemIR::InstId> lhs_inst_ids_;
+  const SemIR::InstId* lhs_inst_ids_it_ = lhs_inst_ids_.end();
+  llvm::ArrayRef<SemIR::InstId> rhs_inst_ids_;
+  const SemIR::InstId* rhs_inst_ids_it_ = rhs_inst_ids_.end();
```
It seems to me that we only need to track (current, end) for each of these, not (begin, current, end). Perhaps we could store just the `ArrayRef`, and slice off the first element when we process it?

danakj replied:
Right, good plan, done.

zygoloid commented on toolchain/check/impl_lookup.cpp:
```diff
@@ -372,23 +382,42 @@ static auto LookupImplWitnessInSelfFacetValue(
       context, context.insts().GetCanonicalLocId(loc_id).ToImplicit(),
-      SemIR::LookupImplWitness{
+      {
           .type_id = GetSingletonType(context, SemIR::WitnessType::TypeInstId),
-          .query_self_inst_id =
-              context.constant_values().GetInstId(query_self_const_id),
+          .query_self_inst_id = canonical_self_inst_id,
           .query_specific_interface_id =
-              context.specific_interfaces().Add(interface),
+              context.specific_interfaces().Add(query_interface),
       });
```
If I'm understanding this correctly, this `AddInstWithoutEval` mechanism is only being used to obtain the `facet_type_id` value that we return below -- the `witness_id` could be computed using `EvalOrAddInst`. And that `facet_type_id` value is always the facet type of self as computed by `LookupImplWitnessInSelfFacetValue`:

```c++
  // Unwrap FacetAccessType without getting the canonical facet value from the
  // self value, as we want to preserve the non-canonical `FacetValue`
  // instruction which can contain the concrete witness.
  query_self_inst_id = UnwrapFacetAccessType(context, query_self_inst_id);
  auto facet_type = context.types().TryGetAs<SemIR::FacetType>(
      context.insts().Get(query_self_inst_id).type_id());
 ```

If that's right, I don't think it's worth the complexity cost we're paying here to preserve this information rather than recomputing it from the `Self` type. (And with the shift in approach I suggested in the main review comment, I don't think we would be directly using this value at all any more.)

danakj replied:
The `AddInstWithoutEval` is used to create the `LookupImplWitness` instruction without evaluating it and causing it to do another impl lookup which we have just done on L389 above. That `LookupImplWitness` instruction is then returned as the witness `InstId`.'

It is the extra pre-work to do what `LookupImplWitness` eval would do on L389 which is used to get the FacetTypeId that the lookup found a witness from (if that's where it found it). That FacetTypeId is always the self type, or None. And that decision on being None or not depends on whether `EvalLookupSingleImplWitness` used the facet type to get the witness.

Does that clarify things or did I misunderstand your meaning?

zygoloid commented on toolchain/check/inst.cpp:
```diff
@@ -193,6 +222,32 @@ auto EvalOrAddInst(Context& context, SemIR::LocIdAndInst loc_id_and_inst)
   }
 }
 
+auto AddInstWithoutEval(Context& context, SemIR::LocIdAndInst loc_id_and_inst)
+    -> SemIR::ConstantId {
+  CARBON_CHECK(!loc_id_and_inst.inst.kind().has_cleanup());
+  switch (loc_id_and_inst.inst.kind().constant_needs_inst_id()) {
+    case SemIR::InstConstantNeedsInstIdKind::No: {
+      return AddImportedConstant(context, loc_id_and_inst.inst);
+    }
+
+    case SemIR::InstConstantNeedsInstIdKind::DuringEvaluation: {
+      // Evaluation temporarily needs an InstId. Add one for now.
+      auto inst_id = AddInstInNoBlockWithOptionalEval(context, loc_id_and_inst,
+                                                      WithoutEval);
+      // TODO: Consider removing `inst_id` from `insts` if it's still the most
+      // recently added instruction.
+      return context.constant_values().Get(inst_id);
+    }
+
+    case SemIR::InstConstantNeedsInstIdKind::Permanent: {
+      // Evaluation needs a permanent InstId. Add the instruction.
+      auto inst_id =
+          AddInstWithOptionalEval(context, loc_id_and_inst, WithoutEval);
+      return context.constant_values().Get(inst_id);
+    }
+  }
+}
```
The semantics of this function seem inconsistent:
* In the first case here, we add it to the constants store.
* In the second case, we add it to no block as an orphan.
* In the third case, we add it to the current block.

It seems like we should probably be providing only a single behavior here (perhaps the first one?), regardless of what evaluation needs, because we're not invoking evaluation.

If we don't pick the third of the above semantics, this function seems to have the wrong name. (The third one is a form of `AddInst`, the second is a form of `AddInstInNoBlock`; for the first thing I might consider something more like `AddReducedConstantWithoutEval`?)

danakj replied:
The only use of this is `DuringEvaluation` and it does use an inst id, but it can use the one referred to by the ConstantId, it doesn't need a further InstId defining the constant, I think.

That said all of this was copied from `EvalOrAddInst`, and the semantics are the same as that except it doesn't do the `Eval` part. We can go with that name and explain what "reduced" means.

zygoloid commented on toolchain/check/inst.h:
```diff
@@ -97,6 +97,23 @@ auto EvalOrAddInst(Context& context, LocT loc, InstT inst)
   return EvalOrAddInst(context, SemIR::LocIdAndInst(loc, inst));
 }
 
+// Add an instruction and return its constant value by canonicalizing it but
+// without evaluating the instruction. The instruction will need to be evaluated
+// later to have any effect.
```
In order for a call to this to be reasonable, I think we also need the caller to have already converted the instruction into a reduced form; that should be documented somehow. (Specifically, I think the constraint that we need here is that the caller knows that evaluating the instruction would only canonicalize the operands but would otherwise produce the same instruction.)

danakj replied:
Yes that is accurate. I guess this is introducing the concept of "reduced". I will add some comments.

zygoloid reviewed (commented):
I think this is doing the right thing for simple cases, but I'm concerned that this approach won't scale up to more complicated examples. For example, if we have `T:! I where .X = C(.Y)`, and the impl witness for type `Z` tells us that `.Y = i32` and `.X = C(i32)`, the comparison we perform will fail.

In order to get this right in general, we could do something along these lines:

* Substitute `.Self` for the type `T` in the requirements.
* Resolve the facet type by performing rewrites into the rewrites. (Either before or after the previous step, I suppose.)
* Check that the resulting LHS and RHS are equal after substitution.

Eg, for `T:! I where .X = C(.Y)`, substituting in `Z` for `.Self` gives `I where Z.(I.X) = C(Z.(I.Y))`, which evaluates to `I where C(i32) = C(i32)`, so the check should succeed.

In order for that kind of approach to work, we'd first need to use rewrite constraints as part of constant-evaluation of `ImplWitnessAccess` instructions. Perhaps we should be looking into that before we go too deep into checking that rewrite constraints are satisfied, so that we can reuse that work as part of the checking process?

danakj committed:
add passing and failing test with `T:! I where .X = C(.Y)`

danakj committed:
remove-unused

danakj committed:
update-comment

danakj committed:
it-before-end

danakj committed:
test-assoc-function-as-rewrite-value

danakj committed:
dont-store-iter-use-arrayref

danakj committed:
AddReducedConstantWithoutEval

danakj commented:
> I think this is doing the right thing for simple cases, but I'm concerned that this approach won't scale up to more complicated examples. For example, if we have `T:! I where .X = C(.Y)`, and the impl witness for type `Z` tells us that `.Y = i32` and `.X = C(i32)`, the comparison we perform will fail.
> 
> In order to get this right in general, we could do something along these lines:
> 
> * Substitute `.Self` for the type `T` in the requirements.
> * Resolve the facet type by performing rewrites into the rewrites. (Either before or after the previous step, I suppose.)
> * Check that the resulting LHS and RHS are equal after substitution.
> 
> Eg, for `T:! I where .X = C(.Y)`, substituting in `Z` for `.Self` gives `I where Z.(I.X) = C(Z.(I.Y))`, which evaluates to `I where C(i32) = C(i32)`, so the check should succeed.
> 
> In order for that kind of approach to work, we'd first need to use rewrite constraints as part of constant-evaluation of `ImplWitnessAccess` instructions. Perhaps we should be looking into that before we go too deep into checking that rewrite constraints are satisfied, so that we can reuse that work as part of the checking process?

I think I get what you're saying here, yeah. I will look into this first (but I have addressed the small stuff in this PR in the meantime and added tests that we'll want to have pass either way).

danakj commented:
> In order for that kind of approach to work, we'd first need to use rewrite constraints as part of constant-evaluation of `ImplWitnessAccess` instructions. Perhaps we should be looking into that before we go too deep into checking that rewrite constraints are satisfied, so that we can reuse that work as part of the checking process?

Given this simple example:

```carbon
interface I { let X:! type; }

fn F(T:! I where .X = ()) {}

fn G(V:! I where .X = ()) {
  F(V);
}
```

I think you're saying that when we construct the specific for `F(V)`, we want to evaluate the `.X` `ImplWitnessAccess` in the signature of `F`.

What we do is we evaluate the `BindSymbolicName` for `T`, which replaces it with the `BindSymbolicName` for `V`. We don't do anything with the `ImplWitnessAccess` in the type of `T` and it's never evaluated at a point where the witness is concrete (an `ImplWitness`).

So then I think what you're saying is that when evaluating `BindSymbolicName` for `T` once we have a `CompileTimeBindValue` for it, we want to construct basically a new witness table by re-evaluating the instructions in the witness table of the `FacetType` of the `BindSymbolicName` for `T` by calling `ReplaceTypeWithConstantValue` . Then comparing the table for each element that is an associated constant with the witness table of the type of the original `BindSymbolicName` for `T`.

Does that sound like what you're thinking?

zygoloid commented:
> Given this simple example:
> 
> ```v
> interface I { let X:! type; }
> 
> fn F(T:! I where .X = ()) {}
> 
> fn G(V:! I where .X = ()) {
>   F(V);
> }
> ```
> 
> I think you're saying that when we construct the specific for `F(V)`, we want to evaluate the `.X` `ImplWitnessAccess` in the signature of `F`.

I think it probably makes sense to do this check as part of the conversion of `V` to the type `I where .X = ()`. (In this case, that's a no-op conversion so maybe we can skip this logic, but of course it wouldn't be in general.)

The left-hand side of the constraint here is `.Self.X` and the right-hand side is `()`. Substituting `.Self` = `V` into both produces an LHS of `V.X` and an RHS of `()`. The LHS is an `ImplWitnessAccess` naming `X` in `V`, and the type of `V` has a rewrite `.X = ()`, so constant evaluation of that `ImplWitnessAccess` should result in `()` (even though the witness is still symbolic) -- that's the new step that we don't have yet, that `ImplWitnessAccess` evaluation should take into account rewrite rules in the type of the left-hand side to directly produce a result even when the impl witness itself is symbolic. That rewrite results in the LHS and RHS evaluating to the same constant, in this case `()`, so the check passes and we can allow the conversion of `V` to type `I where .X = ()`.

danakj commented:
> > Given this simple example:
> > ```v
> > interface I { let X:! type; }
> > 
> > fn F(T:! I where .X = ()) {}
> > 
> > fn G(V:! I where .X = ()) {
> >   F(V);
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I think you're saying that when we construct the specific for `F(V)`, we want to evaluate the `.X` `ImplWitnessAccess` in the signature of `F`.
> 
> I think it probably makes sense to do this check as part of the conversion of `V` to the type `I where .X = ()`. (In this case, that's a no-op conversion so maybe we can skip this logic, but of course it wouldn't be in general.)
> 
> The left-hand side of the constraint here is `.Self.X` and the right-hand side is `()`. Substituting `.Self` = `V` into both produces an LHS of `V.X` and an RHS of `()`. The LHS is an `ImplWitnessAccess` naming `X` in `V`, and the type of `V` has a rewrite `.X = ()`, so constant evaluation of that `ImplWitnessAccess` should result in `()` (even though the witness is still symbolic) -- that's the new step that we don't have yet, that `ImplWitnessAccess` evaluation should take into account rewrite rules in the type of the left-hand side to directly produce a result even when the impl witness itself is symbolic. That rewrite results in the LHS and RHS evaluating to the same constant, in this case `()`, so the check passes and we can allow the conversion of `V` to type `I where .X = ()`.

When you say to substitute the query self type here for `.Self`, we can literally do this with Subst, I guess. Like we did in deduce?

But over there we said that was not ideal, and that we'd like to apply a specific instead. I am trying to wrap my head around how to construct a specific for the target facet type so that when we get constant values we would replace .Self with the query self type. But the FacetType in the query constraint is concrete in this example, even though there's a symbolic value in the rewrite. This is seemingly intentional, introduced in 33110d096cb0c089995dfbf5c1986d4d5c178dad. So we can't replace the .Self in the FacetType through eval. We will _have_ to use Subst I guess.

But is this good?

danakj commented:
https://github.com/carbon-language/carbon-lang/pull/5617 and its successors will be superceding this PR
